staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go

```diff
@@ -17,77 +17,203 @@ limitations under the License.
+	ModeBlocking,
+	ModeBuffered,
+}
+
+const (
 	// FormatLegacy saves event in 1-line text format.
 	FormatLegacy = "legacy"
 	// FormatJson saves event in structured json format.
 	FormatJson = "json"
+
+	// TODO: make this configurable
+	// Number of workers that concurrently write events.
+	workers = 5
+	// Buffer up to 1000 events.
+	defaultBufferSize = 1000
 )
 
+// The plugin name reported in error metrics.
+const pluginName = "log"
+
 // AllowedFormats are the formats known by log backend.
 var AllowedFormats = []string{
 	FormatLegacy,
 	FormatJson,
 }
 
-type backend struct {
+func NewBackend(out io.Writer, format, mode string, groupVersion schema.GroupVersion) (audit.Backend, error) {
+	switch mode {
+	case ModeBuffered:
+		return newBufferedBackend(out, format, groupVersion), nil
+	case ModeBlocking:
+		return newBlockingBackend(out, format, groupVersion), nil
+	default:
+		return nil, fmt.Errorf("log mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+
+}
+
+type bufferedBackend struct {
+	out          io.Writer
+	format       string
+	groupVersion schema.GroupVersion
+	buffer       chan *auditinternal.Event
+	wg           sync.WaitGroup
+}
+
+var _ audit.Backend = &bufferedBackend{}
+
+func newBufferedBackend(out io.Writer, format string, groupVersion schema.GroupVersion) audit.Backend {
+	return &bufferedBackend{
+		out:          out,
+		format:       format,
+		groupVersion: groupVersion,
+		buffer:       make(chan *auditinternal.Event, defaultBufferSize),
+		wg:           sync.WaitGroup{},
+	}
+}
+
+func (b *bufferedBackend) ProcessEvents(events ...*auditinternal.Event) {
+	for i, e := range events {
+		// Per the audit.Backend interface these events are reused after being
+		// sent to the Sink. Deep copy and send the copy to the queue.
+		event := e.DeepCopy()
+
+		// The following mechanism is in place to support the situation when audit
+		// events are still coming after the backend was shut down.
+		var sendErr error
+		func() {
+			// If the backend was shut down and the buffer channel was closed, an
+			// attempt to add an event to it will result in panic that we should
+			// recover from.
+			defer func() {
+				if err := recover(); err != nil {
+					sendErr = errors.New("audit log shut down")
+				}
+			}()
+
+			select {
+			case b.buffer <- event:
+			default:
+				sendErr = errors.New("audit log queue blocked")
+			}
+		}()
+		if sendErr != nil {
+			audit.HandlePluginError(pluginName, sendErr, events[i:]...)
+			return
+		}
+	}
+}
+
+func (b *bufferedBackend) processEvent(ev *auditinternal.Event) {
+	logEvent(b.out, b.format, b.groupVersion, ev)
+}
+
+func (b *bufferedBackend) worker() {
+	for event := range b.buffer {
+		b.processEvent(event)
+	}
+	b.wg.Done()
+}
+
+func (b *bufferedBackend) Run(stopCh <-chan struct{}) error {
+	glog.Infof("Starting audit log backend")
+
+	for i := 0; i < workers; i++ {
+		b.wg.Add(1)
+		go wait.Until(b.worker, time.Second, stopCh)
```

Comment:
If the workers log events in a goroutine couldn't they concurrently write log events to the same io.Writer? E.g. you could end up with overlapping JSON in the output or something. Also io.Writer's aren't guaranteed to be safe for concurrent writes.

Reply:
worker goroutine write events to the same io.Writer, and concurrent writes safety is guaranteed.

Reply:
> concurrent writes safety is guaranteed

Not, it's not. There's nothing about safety in the interface: https://golang.org/src/io/io.go?s=3794:3855#L80

Reply:
Yes, It does not guarantee from interface, but the implement of this os.Stdout or lumberjack.Logger do not have concurrent write issue.

Reply:
It is safe, do not worry.

Reply:
We're dealing with interfaces, not implementation. Implementation might change at any moment

Reply:
A little weird. As you say, the pre-existing code have to add a lock to guarantee concurrent write safety too. Because the current `ProcessEvents` do not have any mutex lock.

Reply:
Current audit log write is concurrent by many http handler goroutines. By the way, lower writer should support concurrent write as file system does.

Reply:
Well, that is actually a bug

> By the way, lower writer should support concurrent write as file system does.

POSIX FS doesn't support atomic writes of size larger than 4K, so it's just a lucky coincidence that all used `io.Writer`s are thread-safe

Reply:
OK, I will do more research， thanks for point out

Reply:
In application level, I do not see any need for extra atomic protect. File system does it for us. And I also not see 4K restriction.

Reply:
This is wrong. Search by `PIPE_BUF`

Reply:
ok